python/lsst/daf/butler/core/dimensions/elements.py

```diff
@@ -0,0 +1,194 @@
+        """
+        return self._primaryKey
+
+    @property
+    def link(self):
+        """Primary key fields that are used only by this dimension, not any
+        dependencies (`frozenset` of `str`).
+        """
+        return self._link
+
+    def dependencies(self, optional=True):
+        """Return the set of dimension units this dimension depends on.
+
+        Parameters
+        ----------
+        optional : `bool`
+            If `True` (default), include optional as well as required
+            dependencies. The primary keys of required dependencies are always
+            included in the primary key of the dimension that depends on them,
+            while the primary keys of optional dependencies are not.
+
+        Returns
+        -------
+        dependencies : `DimensionSet` of `DimensionUnit`
+        """
+        return self._allDependencies if optional else self._requiredDependencies
+
+
+class DimensionUnit(Dimension):
+    r"""A discrete unit of data used to organize `Dataset`\s and associate
+    them with metadata.
+
+    `DimensionUnit` instances represent concepts such as "Instrument",
+    "Detector", "Visit" and "SkyMap", which are usually associated with
+    database tables.  A `DatasetType` is associated with a fixed combination
+    of `DimensionUnit`\s.
+
+    `DimensionUnit` objects are not directly constructable; they can only be
+    obtained from a `DimensionGraph`.
+    """
+
+    def __init__(self, universe, name, config):
+        super().__init__(universe, name, hasRegion=config.get("hasRegion", False), link=config["link"],
+                         required=config.get(".dependencies.required", ()),
+                         optional=config.get(".dependencies.optional", ()),
+                         doc=config["doc"])
+
+    def __repr__(self):
+        return "DimensionUnit({})".format(self.name)
+
+
+class DimensionJoin(Dimension):
+    r"""A join that relates two or more `DimensionUnit`\s.
+
+    `DimensionJoin`\s usually map to many-to-many join tables or views that
+    relate `DimensionUnit` tables.
+
+    `DimensionJoin` objects are not directly constructable; they can only be
+    obtained from a `DimensionGraph`.
+    """
+
+    def __init__(self, universe, name, config):
+        lhs = list(config["lhs"])
+        rhs = list(config["rhs"])
+        super().__init__(universe, name, hasRegion=config.get("hasRegion", False), link=(),
+                         required=lhs + rhs, optional=(), doc=config["doc"])
+        self._lhs = DimensionUnitSet(universe, lhs, optional=False)
+        self._rhs = DimensionUnitSet(universe, rhs, optional=False)
+        # self._summarizes initialized later in DimensionMeta.load.
+
+    @property
+    def lhs(self):
+        r"""The `DimensionUnits`\s on the left hand side of the join
+        (`DimensionUnitSet`).
+
+        Left vs. right is completely arbitrary; the terminology simply provides
+        an easy way to distiguish between the two sides.
```

Comment:
Should we drop lhs/rhs entirely if they are not useful and just do `def joins(): return (lhs, rhs)`?